manifests/init.pp

```diff
@@ -1,22 +1,85 @@
-# Class: sftp_jail
-# ===========================
+# @summary sftp_jail class
 #
-# Full description of class sftp_jail here.
+# @param chroot_base
+#   The directories path, which holds all jails.
 #
-# Parameters
-# ----------
+# @param sftp_users
+#   See `Sftp_jail::Sftp_Users` and `Sftp_jail::Sftp_User` types.
 #
-# * `sample parameter`
-#   Explanation of what this parameter affects and what it defaults to.
-#   e.g. "Specify one or more upstream ntp servers as an array."
+# @param manage_users
+#   Should this module manage the user resources?
+#
+# @param sub_dirs
+#   Default subdirectories to ensure in every users home. Can be overwritten for
+#   each user seperatly.
+#
+# @param merge_subdirs
+#   Merge sub_dirs with the default sub_dirs?
+#
+# @param admin_sub_dirs
+#   Default subdirectories to ensure in every admins home. Can be overwritten for
+#   each admin seperatly.
+#
+# @param password_authentication
+#   Password Authentication setting for SFTP users and Admins.
 #
 class sftp_jail (
-  Stdlib::Absolutepath $chroot_base = '/chroot',
+  Sftp_jail::Sftp_Users    $sftp_users              = {},
+  Stdlib::Absolutepath     $chroot_base             = '/chroot',
+  Boolean                  $manage_users            = false,
+  Sftp_jail::Sub_dirs      $sub_dirs                = [],
+  Boolean                  $merge_subdirs           = false,
+  Sftp_jail::Sub_dirs      $admin_sub_dirs          = [],
+  Sftp_jail::Pw_auth       $password_authentication = {},
 ) {
+  $sftp_admins = $sftp_users.filter |$k, $v| { $v['sftp_admin'] }
+
   file { $chroot_base:
     ensure => 'directory',
     owner  => 'root',
     group  => 'root',
     mode   => '0755',
   }
+
+  unless $sftp_admins.empty {
+    file { "${chroot_base}/home":
+      ensure  => 'directory',
+      owner   => 'root',
+      group   => 'root',
+      mode    => '0755',
+      require => File[$chroot_base],
+    }
+    $sftp_user_groups = ($sftp_users - $sftp_admins).filter |$k, $v| {
+      $v['ensure'] == 'present'
+    }.map |$k, $v| {
+      if $v['group'] {
+        $v['group']
+      } else {
+        $k
+      }
+    }
+  }
+
+  $sftp_users.each |$k, $v| {
+    if $v['sftp_admin'] {
+      $jail_base = $chroot_base
+      $sub_dirs = $admin_sub_dirs
+      $merge_subdirs = false
+    }else{
+      $jail_base = undef
+    }
+
+    sftp_jail::jail { default:
+        sub_dirs      => $sub_dirs,
+        merge_subdirs => $merge_subdirs,
+      ;
+      $k:
+        *                       => $v,
+        jail_base               => $jail_base,
+        groups                  => $sftp_user_groups,
+        manage_user             => $manage_users,
+        password_authentication => $password_authentication['sftp_users'],
+      ;
+    }
```

Comment:
How about:
```suggestion
    if $v['sftp_admin'] {
      $jail_base = $chroot_base
      $options = {
        'sub_dirs' => $admin_sub_dirs,
        'merge_subdirs' => false,
      }
    } else {
      $jail_base = undef
      $options = {}
    }

    sftp_jail::jail { $k:
        jail_base               => $jail_base,
        groups                  => $sftp_user_groups,
        manage_user             => $manage_users,
        password_authentication => $password_authentication['sftp_users'],
        *                       => $options + $v,
    }
```

Reading this, it also makes me wonder if password_authentication is correct here. It doesn't take into account the user vs admin?

Reply:
Accepted your change and fixed password auth: https://github.com/voxpupuli/puppet-sftp_jail/pull/69/commits/645b080e99f96583f8e05aac060f6682d20c20a3#diff-c6c3a824d185ff01e063011517a1b13a604dc3d7b63d1c67a30451c10df02bcbR68-R81

Reply:
Default values got lost, fixed in: https://github.com/voxpupuli/puppet-sftp_jail/pull/69/commits/45e1b6fcf925f20752590333552ba8174e6697f7